app/models/user.rb

```diff
@@ -117,10 +119,59 @@ def <=>(other)
     login.downcase <=> other.login.downcase
   end
 
+  def readable_by?(user)
+    user.administrator? ||
+        self.id == user.id ||
+        visible_to_teacher?(user) ||
+        visible_to_assistant?(user)
+  end
+
+  def visible_to_teacher?(teacher)
+    teacher.organizations.each do |org|
+      return true if self.student_in_organization?(org)
+    end
+    false
+  end
+
+  def visible_to_assistant?(assistant)
+    assistant.assisted_courses.each do |c|
+      return true if self.student_in_course?(c)
```

Comment:
As discusses bit more in irc, simplified this part a bit, but visibility stays bit heavy operation for now, and shouldn't be used in looping for views' list generation. Relevant comments added.